From a1ff9f0decd722dfbd5f00bba04984283a345758 Mon Sep 17 00:00:00 2001
From: Omri Itach <omrii@marvell.com>
Date: Tue, 31 May 2016 10:03:39 +0300
Subject: [PATCH 1181/1240] fix: utmi: a80x0: fix duplicated dedicated phys
 initialization

comphy_dedicated_phys_init was called twice in cp110_init (once per CP).
Fixed this to be called only once from comphy_core init.

UTMI should be independant from the comphy driver,
since it can set all UTMI interfaces, regardless of
the comphy lanes status.

Change-Id: I431d849b122ef0d6c96758c5366a71157a689a4c
Signed-off-by: Omri Itach <omrii@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/30185
Tested-by: Star_Automation <star@marvell.com>
Reviewed-by: Hanna Hawa <hannah@marvell.com>
Reviewed-by: Haim Boot <hayim@marvell.com>
---
 arch/arm/include/asm/arch-mvebu/comphy.h |  1 +
 drivers/phy/comphy_a3700.c               |  4 +---
 drivers/phy/comphy_core.c                | 11 ++++++++---
 drivers/phy/comphy_cp110.c               |  5 +----
 4 files changed, 11 insertions(+), 10 deletions(-)

diff --git a/arch/arm/include/asm/arch-mvebu/comphy.h b/arch/arm/include/asm/arch-mvebu/comphy.h
index 51e66c7..1ed1e4a 100644
--- a/arch/arm/include/asm/arch-mvebu/comphy.h
+++ b/arch/arm/include/asm/arch-mvebu/comphy.h
@@ -94,6 +94,7 @@ void reg_set16(void __iomem *addr, u16 data, u16 mask);
 void reg_set_silent16(void __iomem *addr, u16 data, u16 mask);
 u32 comphy_init(const void *blob);
 u32 polling_with_timeout(void __iomem *addr, u32 val, u32 mask, unsigned long usec_timout);
+void comphy_dedicated_phys_init(void);
 
 #ifdef CONFIG_TARGET_ARMADA_38X
 int comphy_a38x_init(struct chip_serdes_phy_config *ptr_chip_cfg, struct comphy_map *comphy_map_data);
diff --git a/drivers/phy/comphy_a3700.c b/drivers/phy/comphy_a3700.c
index ad46431..4a0ad2a 100644
--- a/drivers/phy/comphy_a3700.c
+++ b/drivers/phy/comphy_a3700.c
@@ -748,7 +748,7 @@ static int comphy_sgmii_power_up(u32 lane, u32 speed, u32 invert)
 /***************************************************************************************************
   * comphy_dedicated_phys_init
  ***************************************************************************************************/
-static void comphy_dedicated_phys_init(void)
+void comphy_dedicated_phys_init(void)
 {
 	int node, count, usb32, ret = 1;
 	const void *blob = gd->fdt_blob;
@@ -865,8 +865,6 @@ int comphy_a3700_init(struct chip_serdes_phy_config *ptr_chip_cfg, struct comphy
 			error("PLL is not locked - Failed to initialize lane %d\n", lane);
 	}
 
-	comphy_dedicated_phys_init();
-
 	debug_exit();
 	return ret;
 }
diff --git a/drivers/phy/comphy_core.c b/drivers/phy/comphy_core.c
index d006f4f..b2dd0d5 100644
--- a/drivers/phy/comphy_core.c
+++ b/drivers/phy/comphy_core.c
@@ -141,9 +141,6 @@ u32 comphy_init(const void *blob)
 	chip_count = fdtdec_find_aliases_for_id(blob, "comphy",
 			COMPAT_MVEBU_COMPHY, comphy_list, COMPHY_MAX_CHIP);
 
-	if (chip_count <= 0)
-		return 1;
-
 	for (i = 0; i < chip_count ; i++) {
 		if (chip_count > 1)
 			printf("Comphy chip #%d:\n", i);
@@ -209,8 +206,16 @@ u32 comphy_init(const void *blob)
 		/* PHY print SerDes status */
 		comphy_print(ptr_chip_cfg, comphy_map_data);
 	}
+
+	/* Initialize dedicated PHYs (not muxed SerDes lanes) */
+	comphy_dedicated_phys_init();
+
 	debug_exit();
 
+	/*  Return failure in case no comphy unit was set in Device tree */
+	if (chip_count <= 0)
+		return 1;
+
 	return 0;
 }
 
diff --git a/drivers/phy/comphy_cp110.c b/drivers/phy/comphy_cp110.c
index 83cea45..cd60acc 100644
--- a/drivers/phy/comphy_cp110.c
+++ b/drivers/phy/comphy_cp110.c
@@ -1359,7 +1359,7 @@ static void comphy_utmi_phy_init(u32 utmi_phy_count, struct utmi_phy_data *cp110
 
 /* comphy_dedicated_phys_init initialize the dedicated PHYs - not muxed SerDes lanes
 ** e.g. UTMI PHY */
-static void comphy_dedicated_phys_init(void)
+void comphy_dedicated_phys_init(void)
 {
 	struct utmi_phy_data cp110_utmi_data[MAX_UTMI_PHY_COUNT];
 	int node_list[MAX_UTMI_PHY_COUNT], node;
@@ -1521,9 +1521,6 @@ int comphy_cp110_init(struct chip_serdes_phy_config *ptr_chip_cfg, struct comphy
 			error("PLL is not locked - Failed to initialize lane %d\n", lane);
 	}
 
-	/* Initialize dedicated PHYs (not muxed SerDes lanes) */
-	comphy_dedicated_phys_init();
-
 	debug_exit();
 	return 0;
 }
-- 
1.9.1

